### DISCUSSION TITLE
How is processor utilization calculated for each process?

### DISCUSSION THREAD

**Discussion by KovalikVadimdev** [winsiderss/systeminformer#2239] [Questions / Answers]
I've looked through all the folders and files, but I can't find where the CPU usage by each process is calculated.
Please write how to find the CPU usage by each process or point to the code where it can be seen.
I will be very grateful for this.

**Comment by Advait251206** [comment]
If you're looking for the exact implementation, I'd start by searching for references to:

```c
NtQuerySystemInformation(
    SystemProcessInformation,
    ...
)
```

and the process provider code that periodically refreshes process statistics.

System Informer doesn't typically calculate CPU usage from a single snapshot. Instead, it samples process CPU times over an interval and computes utilization from the delta between two measurements.

Conceptually, the calculation looks something like:

```text
ProcessCpuUsage =
    (ΔKernelTime + ΔUserTime)
    /
    ΔElapsedTime
```

and then normalized by the number of logical processors:

```text
CPU% =
    100 *
    (ΔKernelTime + ΔUserTime)
    /
    (ΔElapsedTime * ProcessorCount)
```

The inputs usually come from fields exposed through `SYSTEM_PROCESS_INFORMATION`, such as:

- `KernelTime`
- `UserTime`

which are accumulated CPU times since process creation.

The workflow is generally:

1. Take a snapshot of process times.
2. Wait for the refresh interval.
3. Take another snapshot.
4. Compute deltas.
5. Convert the result to a percentage.

For the source code, I'd suggest looking at:

- Process provider code (`procprv*`)
- System information provider code
- Functions related to process statistics updates
- Any routines containing names such as:
  - `CpuUsage`
  - `CpuKernelUsage`
  - `CpuUserUsage`
  - `UpdateProcessStatistics`
  - `ProcessProviderUpdate`

One thing to note is that System Informer may use a slightly different normalization method than Task Manager depending on whether it's displaying:

- Per-core utilization,
- Total system utilization,
- Or percentages normalized across all CPUs.

If you're modifying the codebase, it may also help to search for where the `CPU` column is populated, since the column update path will eventually lead back to the utilization calculation logic.
